### PR TITLE
Histogram CT Zero ingestion

### LINKED ISSUE #13384: tsdb: need histogram support for created-timestamp handling
### Proposal

OM introduced created timestamps, but doesn't support native histograms yet. However, we have backported created-timestamps to the classic protobuf format, which also supports native histograms. (And of course, OM is supposed to support native histograms soon.)

Combining both features, we run into the problem that [AppendCTZeroSample](https://github.com/prometheus/prometheus/blob/6cd24d87b0d4e3acd4503b061d32c5aa6d8883a9/tsdb/head_append.go#L389) will always insert a float zero sample, even into a series of histograms. Which will make rate calculation impossible whenever the zero float sample and any histogram sample is in the range at the same time.

So we need to find a way to insert a zero histogram in case the created timestamp comes with a histogram sample.

### PR DESCRIPTION
**Pull request by ArthurSens**
Fixes #13384

Fundamentally it follows the same logic used in `AppendCTZeroSample()`, used to append zeros for usual samples, but for histograms, we're using Go's zero value for `histogram.Histogram{}` and `histogram.FloatHistogram{}`

The tricky thing here is that the Histogram format has CounterResetHints, which the query engine uses to detect resets. From my understanding, however, this is a no-op during the scrape and the storage layer handles the addition of counter-reset hints. This PR is only tackling the scrape part of things.

There is one test in head_appender_test that queries the appended data and counter-reset hints are verified as well

### CODE REVIEW

File: tsdb/head_append.go [hunk 1]
```diff
@@ -673,6 +683,89 @@ func (a *headAppender) AppendHistogram(ref storage.SeriesRef, lset labels.Labels
...
+		zeroFloatHistogram := &histogram.FloatHistogram{}
+		s.Lock()
+		if err := s.appendableFloatHistogram(ct, zeroFloatHistogram); err != nil {
+			s.Unlock()
+			if errors.Is(err, storage.ErrOutOfOrderSample) {
+				return 0, storage.ErrOutOfOrderCT
+			}
+		}
+		s.pendingCommit = true
+		s.Unlock()
```
[the remark this answers is not in the document]
Comment (ArthurSens) [reply tsdb/head_append.go:754]: This part is basically a copy of AppendHistogram, which does the same thing. I'd be surprised if that's the problem

https://github.com/prometheus/prometheus/blob/95b3c49836acc3212a98016daa21b5a199d4f6c8/tsdb/head_append.go#L650-L657

**Comment by ArthurSens** [comment]
Alright, only 1 test is failing now: `TestHeadAppender_AppendHistogramCTZeroSample`.

What this test does is ingest Histogram samples with created timestamps in several different scenarios. Some of those scenarios should reject the sample and others should ingested. After that, we query the ingested samples and verify that a histogram zero sample and the normal histogram sample are retrieved as expected.

The test is failing right now because the histogram zero samples are being modified after ingestion with a non-zero `CounterResetHint`. @beorn7 @krajorama, I'm not familiar with this part. Is it expected that histograms are modified during ingestion? Should I ignore this field when comparing the ingested/queried histogram?

**Comment by Maniktherana** [comment]
> Alright, only 1 test is failing now: `TestHeadAppender_AppendHistogramCTZeroSample`.
> 
> What this test does is ingest Histogram samples with created timestamps in several different scenarios. Some of those scenarios should reject the sample and others should ingested. After that, we query the ingested samples and verify that a histogram zero sample and the normal histogram sample are retrieved as expected.
> 
> The test is failing right now because the histogram zero samples are being modified after ingestion with a non-zero `CounterResetHint`. @beorn7 @krajorama, I'm not familiar with this part. Is it expected that histograms are modified during ingestion? Should I ignore this field when comparing the ingested/queried histogram?

relevant @bwplotka https://github.com/prometheus/prometheus/issues/13384#issuecomment-2306538815

**Comment by krajorama** [comment]
> Alright, only 1 test is failing now: `TestHeadAppender_AppendHistogramCTZeroSample`.
> 
> What this test does is ingest Histogram samples with created timestamps in several different scenarios. Some of those scenarios should reject the sample and others should ingested. After that, we query the ingested samples and verify that a histogram zero sample and the normal histogram sample are retrieved as expected.
> 
> The test is failing right now because the histogram zero samples are being modified after ingestion with a non-zero `CounterResetHint`. @beorn7 @krajorama, I'm not familiar with this part. Is it expected that histograms are modified during ingestion? Should I ignore this field when comparing the ingested/queried histogram?

Yes, the appended histogram as well as the resulting histogram in TSDB can be modified by Append.
The counter reset hint is set to help the PromQL engine in counter reset detection. If the count reset hint is left 0, then the PromQL engine runs the detection every time , which is super costly for native histograms, so the Appender tries to set the counter reset to either 1 or 2 (detected reset, detected no reset) to help PromQL. So reading back different hint is totally expected. Also if I understand correctly, the zero histogram is getting written into TSDB as an extra sample. That **will** induce a counter reset, so to see Counter resets is totally expected.

There's two strategies in tests around this:
1. If we are not interested in the counter resets, in which case you'd set the counter reset hint to 0 in the actual samples before making the comparison with the expected samples.
2. If we are interested in the counter resets, in which case you have to know/calculate what the expected hint should be.

If `AppendHistogramCTZeroSample` was just simply calling `AppendHistogram` I would say go with 1. , but it's more complicated and might get even more complicated due to out-of-order (#14546 ) so I think we need to go for 2 and encode the expected counter resets.

**Review by krajorama — COMMENTED** [review]
first pass feedback on tests

**Comment by krajorama** [comment]
> > Appender tries to set the counter reset to either 1 or 2 (detected reset, detected no reset) to help PromQL.
> 
> > Also if I understand correctly, the zero histogram is getting written into TSDB as an extra sample. That will induce a counter reset, so to see Counter resets is totally expected.
> 
> Ah ok that makes sense then. Yes, we should always see a CounterResetHint==2 when ingesting a zero sample histogram.
> 
> But that leaves me with another question: If we're scraping a histogram for the first time and this histogram has set CT during exposition. Should the zero sample also contain ResetHint==2? Theoretically there was no reset since it's the first sample ever ingested for the series

Yep, you got it. If there's no previous sample, we encode 0 (unknown reset) because later when the chunk is read back there might be some previous sample loaded from disk for example and we cannot say counter reset or no reset with any confidence, so we encode "unknown".

File: tsdb/head_append.go [hunk 2]
```diff
@@ -673,6 +683,89 @@ func (a *headAppender) AppendHistogram(ref storage.SeriesRef, lset labels.Labels
...
+		if l, dup := lset.HasDuplicateLabelNames(); dup {
+			return 0, fmt.Errorf(`label name "%s" is not unique: %w`, l, ErrInvalidSample)
+		}
+
+		var created bool
+		var err error
+		s, created, err = a.head.getOrCreate(lset.Hash(), lset)
+		if err != nil {
+			return 0, err
+		}
```
Comment (krajorama) [inline tsdb/head_append.go:711]: This is different from what `AppendCTZeroSample` does, I suspect because `getOrCreate(lset)` doesn't return `created` ? Let's just add created as a new return value and consolidate.

```
func (a *headAppender) getOrCreate(lset labels.Labels) (*memSeries, error) {
```

File: tsdb/head_append.go [hunk 3]
```diff
@@ -673,6 +683,89 @@ func (a *headAppender) AppendHistogram(ref storage.SeriesRef, lset labels.Labels
...
+		if created {
+			switch {
+			case h != nil:
+				s.lastHistogramValue = &histogram.Histogram{}
+			case fh != nil:
+				s.lastFloatHistogramValue = &histogram.FloatHistogram{}
+			}
```
Comment (krajorama) [inline tsdb/head_append.go:718]: Let's do this after we ran s.appendableHistogram/s.appendableFloatHistogram successfully, under lock.

File: tsdb/head_append.go (hunk 2 above)
Comment (krajorama) [reply tsdb/head_append.go:711]: maybe don't even need the return value if https://github.com/prometheus/prometheus/pull/14694#discussion_r1734563712 is done

Comment (krajorama) [reply tsdb/head_append.go:711]: as discussed IRL, this comes from AppendHistogram, fixing this would be a different issue

**Comment by ArthurSens** [comment]
Converted to draft while I work on the failing test in head_appender

File: tsdb/head_test.go [hunk 4]
```diff
@@ -6038,6 +6038,160 @@ func TestHeadAppender_AppendCTZeroSample(t *testing.T) {
...
+		appender := head.Appender(context.Background())
+		for _, sample := range tc.appendableHistograms {
+			ref, err := appender.AppendHistogramCTZeroSample(0, lbls, sample.ts, sample.ct, sample.h, sample.fh)
+			require.NoError(t, err)
+			_, err = appender.AppendHistogram(ref, lbls, sample.ts, sample.h, sample.fh)
+			require.NoError(t, err)
+		}
+		require.NoError(t, appender.Commit())
+
+		q, err := NewBlockQuerier(head, math.MinInt64, math.MaxInt64)
```
[the remark this answers is not in the document]
Comment (ArthurSens) [reply tsdb/head_test.go]: Thanks for the hint!

**Comment by ArthurSens** [comment]
This PR will conflict with #14546 once merged. I don't mind resolving the conflicts here, so it's up to you which PR gets merged first @krajorama!

**Comment by ArthurSens** [comment]
Converting to draft as I work through the conflicts

**Comment by krajorama** [comment]
> Converting to draft as I work through the conflicts

Many thanks for taking the hit, the OOO NH PR was months in the making.

**Comment by ArthurSens** [comment]
> Many thanks for taking the hit, the OOO NH PR was months in the making.

No problems :)
PR should be ready for review now

File: tsdb/head_append.go [hunk 5]
```diff
@@ -747,6 +757,91 @@ func (a *headAppender) AppendHistogram(ref storage.SeriesRef, lset labels.Labels
...
+		if l, dup := lset.HasDuplicateLabelNames(); dup {
+			return 0, fmt.Errorf(`label name "%s" is not unique: %w`, l, ErrInvalidSample)
+		}
+
+		var created bool
+		var err error
+		s, created, err = a.head.getOrCreate(lset.Hash(), lset)
+		if err != nil {
+			return 0, err
+		}
```
Comment (ArthurSens) [inline tsdb/head_append.go:785]: This part can be abstracted into a separate function, just like we do when appending floats, and we reduce duplicated code with `AppendHistogram`. I'd like to do it in a separate PR though

File: scrape/manager_test.go [hunk 6]
```diff
@@ -858,6 +860,178 @@ func TestManagerCTZeroIngestion(t *testing.T) {
...
+					ScrapeProtocols: []config.ScrapeProtocol{config.PrometheusProto},
+				},
+				ScrapeConfigs: []*config.ScrapeConfig{{JobName: "test"}},
+			}))
+
+			once := sync.Once{}
+			// Start fake HTTP target to that allow one scrape only.
+			server := httptest.NewServer(
+				http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+					fail := true // TODO(bwplotka): Kill or use?
```
Comment (jesusvazquez) [inline scrape/manager_test.go:958]: Do we need the TODO?

File: tsdb/head_append.go [hunk 7]
```diff
@@ -747,6 +757,91 @@ func (a *headAppender) AppendHistogram(ref storage.SeriesRef, lset labels.Labels
...
+				Labels: lset,
+			})
+		}
+	}
+
+	switch {
+	case h != nil:
+		zeroHistogram := &histogram.Histogram{}
+		s.Lock()
+		_, _, err := s.appendableHistogram(ct, zeroHistogram, a.headMaxt, a.minValidTime, a.oooTimeWindow, false) // OOO is not allowed for CTZeroSamples.
```
Comment (jesusvazquez) [inline tsdb/head_append.go:804]: Why OOO is not allowed for CT Zero Samples?

**Review by jesusvazquez — COMMENTED** [review]
Looks good! Left one question to fully understand.

File: tsdb/head_append.go (hunk 7 above)
Comment (ArthurSens) [reply tsdb/head_append.go:804]: That's because after the first sample, every CT is OOO. I'm not sure if WBL accepts samples with duplicated timestamps, but it doesn't sound efficient to append the same 0 sample with the same timestamp on every scrape 🤔

Comment (ArthurSens) [reply tsdb/head_append.go:804]: That's something that we can probably improve and eventually allow OOO CT. 

I guess I could improve the comment here with a better explanation of why OOO CT is not allowed TODAY, if my comment doesn't make sense in the future then people might feel comfortable in changing the codebase 🤔

Comment (krajorama) [inline tsdb/head_append.go:804]: Once the OOO native histograms feature is stable , that boolean at the end `oooHistogramsEnabled` will go away, so you also need to handle the case when the function returns `isOOO==true` like the float case!

File: tsdb/head_append.go [hunk 8]
```diff
@@ -747,6 +757,91 @@ func (a *headAppender) AppendHistogram(ref storage.SeriesRef, lset labels.Labels
...
+		a.histograms = append(a.histograms, record.RefHistogramSample{
+			Ref: s.ref,
+			T:   ct,
+			H:   zeroHistogram,
+		})
+		a.histogramSeries = append(a.histogramSeries, s)
+	case fh != nil:
+		zeroFloatHistogram := &histogram.FloatHistogram{}
+		s.Lock()
+		_, _, err := s.appendableFloatHistogram(ct, zeroFloatHistogram, a.headMaxt, a.minValidTime, a.oooTimeWindow, false) // OOO is not allowed for CTZeroSamples.
```
Comment (krajorama) [inline tsdb/head_append.go:822]: same here, handle return value `isOOO` as the boolean flag `false` will go away

File: tsdb/head_append.go (hunk 7 above)
Comment (krajorama) [reply tsdb/head_append.go:804]: In fact I'd call the function with true, so we just handle the same way as floats.

File: tsdb/head_test.go [hunk 9]
```diff
@@ -6363,6 +6363,154 @@ func TestHeadAppender_AppendCTZeroSample(t *testing.T) {
 	}
 }
 
+func TestHeadAppender_AppendHistogramCTZeroSample(t *testing.T) {
```
Comment (krajorama) [inline tsdb/head_test.go:6366]: Knowing that OOO native histograms will have more changes coming up I think we need a regression test to cover the case of ct falling behind the last ingested sample:
```
{ts:100, h: testHistogram, ct: 1}
{ts:101, h: testHistogram, ct: 1}
```
I'm surprised that there's no such test for floats .

Comment (krajorama) [inline tsdb/head_test.go:6366]: nit: In general we've tried to keep tests more generic, see for example `TestWALSampleAndExemplarOrder` or even `TestHeadMinOOOTimeUpdate`. So we keep the testcases in sync.

File: tsdb/head_append.go (hunk 7 above)
Comment (ArthurSens) [reply tsdb/head_append.go:804]: Ah cool, thanks for the heads up! I'm adjusting the PR

File: tsdb/head_test.go (hunk 9 above)
Comment (ArthurSens) [reply tsdb/head_test.go:6366]: Not sure what is being requested here 🤔. Could you elaborate?

Comment (ArthurSens) [reply tsdb/head_test.go:6366]: I've just added a new sample to the first test case :)

File: scrape/manager_test.go (hunk 6 above)
Comment (ArthurSens) [reply scrape/manager_test.go:958]: Probably not, but @Maniktherana is reworking this test case in https://github.com/prometheus/prometheus/pull/14738. I'm hoping to rebase this PR if that one is merged first, or just open a new pull request updating this one if his PR takes too long. Would that be ok?

File: tsdb/head_test.go (hunk 9 above)
Comment (krajorama) [reply tsdb/head_test.go:6366]: It's easier to keep `TestHeadAppender_AppendHistogramCTZeroSample` and `TestHeadAppender_AppendCTZeroSample` in sync interms of testcases if they are merged together and the types (float, integer histogram, float histogram) are just sub-testcases like in the tests a I mentioned. The idea is to avoid someone finding a bug in floats, only updating the float test, but missing that there are histograms...

Comment (ArthurSens) [reply tsdb/head_test.go]: Got it, and good idea! I've just pushed a commit with this change

**Comment by ArthurSens** [comment]
It should be ready for another review

File: tsdb/head_append.go [hunk 10]
```diff
@@ -747,6 +757,107 @@ func (a *headAppender) AppendHistogram(ref storage.SeriesRef, lset labels.Labels
...
+			T:   ct,
+			FH:  zeroFloatHistogram,
+		})
+		a.floatHistogramSeries = append(a.floatHistogramSeries, s)
+	}
+
+	if ct > a.maxt {
+		a.maxt = ct
+	}
+	return 0, nil
```
Comment (krajorama) [inline tsdb/head_append.go:858]: Inconsistency with float handling:

```suggestion
	return storage.SeriesRef(s.ref), nil
```

File: tsdb/head_test.go [hunk 11]
```diff
@@ -6348,10 +6476,21 @@ func TestHeadAppender_AppendCTZeroSample(t *testing.T) {
...
+
+				// Append histograms if it's a histogram test case
+				if sample.h != nil || sample.fh != nil {
+					ref, err := a.AppendHistogramCTZeroSample(0, lbls, sample.ts, sample.ct, sample.h, sample.fh)
+					require.NoError(t, err)
+					_, err = a.AppendHistogram(ref, lbls, sample.ts, sample.h, sample.fh)
+					require.NoError(t, err)
+				}
 			}
 			require.NoError(t, a.Commit())
```
Comment (krajorama) [inline tsdb/head_test.go:6495]: nit: it's kind of weird that we don't test with multiple Commits() , thus the path to return out of order is never exercised. As far as I can tell that was the case for floats anyway, so maybe another PR?

Comment (ArthurSens) [reply tsdb/head_test.go:6495]: Interesting. Yeah, we haven't tested OOO well enough, that was a good call out. If that's ok I'd be happy to extend the tests here to cover OOO in another PR

File: tsdb/head_append.go (hunk 10 above)
Comment (ArthurSens) [reply tsdb/head_append.go:858]: Good catch! It's not only inconsistent, it would be incorrect to always return 0! The reference is used in following Append calls

**Comment by ArthurSens** [comment]
Just did a quick experiment with an app exposing native histograms, I can see the first one being ingested as a 0 🥳 

image

**Review by krajorama — APPROVED** [review]
LGTM

**Comment by ArthurSens** [comment]
Taking the liberty to merge!
